### PR TITLE
FreeObject: hide harder from the compiler, tweak types

### PR DESCRIPTION
**Pull request by nwf**
On CHERI, the compiler will always issue a warning for
`reinterpret_cast<T*>(vaddr_t)` and similar expressions, and of course, if the
compiler can see far enough into the types, the presence of `if constexpr` will
not save us.  Therefore, lift the conditional out to two definitions of
`FreeObject::encode` using `std::enable_if_t` to gate which is used.

While in the neighborhood, use `address_t` rather than `uintptr_t`; the former
is the latter on all non-CHERI platforms, but is `vaddr_t` rather than
`uintcap_t` on CHERI.

### CODE REVIEW

**Comment by jrtc27** [comment]
Perhaps you want to move to ptraddr_t from vaddr_t? The latter is a deprecated CheriBSDism.

**Comment by jrtc27** [comment]
Oh, you don't actually use vaddr_t anywhere in snmalloc, just the commit message here.

**Comment by nwf** [comment]
Oh, but good call.  The `address_t = vaddr_t` bit is still pending in my CHERIfication tree and it'd be best to fix that before I forget.

**Comment by mjp41** [comment]
I think this is going to have some interesting interactions with #304.

File: src/mem/freelist.h [hunk 1]
```diff
@@ -59,29 +59,44 @@ namespace snmalloc
   {
     FreeObject* next_object;
 
-    static FreeObject* encode(uintptr_t local_key, FreeObject* next_object)
+    /**
+     * On architectures which use IntegerPointers, we can obfuscate our free
+     * lists and use this to drive some probabilistic checks for integrity
+     */
+    template<typename AAL = Aal, typename T = FreeObject>
+    static std::enable_if_t<aal_supports<IntegerPointers, AAL>, T*>
```
Comment (mjp41) [inline src/mem/freelist.h:67]: Can you make this
```
static std::enable_if_t<aal_supports<IntegerPointers, AAL> && CHECK_CLIENT, T*>
```
?

Then you can remove the `ifdef` inside, and use the second encode for that case.

**Review by mjp41 — CHANGES_REQUESTED** [review]
I  changed the `uintptr_t` to address in another PR, as I was touching this already.  I think it is just the hiding left.

**Review by mjp41 — APPROVED** [review]
LGTM, just a little conflict with my merged changes.

**Review by mjp41 — APPROVED** [review]

**Pull request merged**